src/backend/distributed/planner/recursive_planning.c

```diff
@@ -316,6 +302,211 @@ RecursivelyPlanCTEs(Query *query, RecursivePlanningContext *planningContext)
 }
 
 
+/*
+ * RecursivelyPlanSubqueryWalker recursively finds all the Query nodes and
+ * recursively plans if necessary.
+ */
+static bool
+RecursivelyPlanSubqueryWalker(Node *node, RecursivePlanningContext *context)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, Query))
+	{
+		Query *query = (Query *) node;
+		DeferredErrorMessage *error = NULL;
+
+		context->level += 1;
+		error = RecursivelyPlanSubqueriesAndCTEs(query, context);
+		if (error != NULL)
+		{
+			RaiseDeferredError(error, ERROR);
+		}
+		context->level -= 1;
+
+		if (ShouldRecursivelyPlanSubquery(query, context))
+		{
+			RecursivelyPlanSubquery(query, context);
+		}
+
+		return false;
+	}
+
+	return expression_tree_walker(node, RecursivelyPlanSubqueryWalker, context);
+}
+
+
+/*
+ * ShouldRecursivelyPlanSubquery decides whether the input subquery should be recursively
+ * planned or not.
+ *
+ * For the details, see the cases in the function.
+ */
+static bool
+ShouldRecursivelyPlanSubquery(Query *subquery, RecursivePlanningContext *context)
+{
+	bool shouldRecursivelyPlan = false;
+
+	if (FindNodeCheckInRangeTableList(subquery->rtable, IsLocalTableRTE))
+	{
+		/*
+		 * Postgres can always plan queries that don't require distributed planning.
+		 * Note that we need to check this first, otherwise the calls to the many other
+		 * Citus planner functions would error our due to local relations.
+		 */
+		shouldRecursivelyPlan = true;
+	}
+	else
+	{
+		if (DeferErrorIfCannotPushdownSubquery(subquery, false) == NULL)
+		{
+			/*
+			 * Citus can pushdown this subquery, no need to recursively
+			 * plan which is much expensive than pushdown.
+			 */
+			shouldRecursivelyPlan = false;
+		}
+		else if (TaskExecutorType == MULTI_EXECUTOR_TASK_TRACKER &&
+				 SingleRelationRepartitionSubquery(subquery))
+		{
+			/* Citus could plan this subquery through re-partitioning */
+			shouldRecursivelyPlan = false;
+		}
+		else if (DeferErrorIfQueryNotSupported(subquery) == NULL)
```

Comment:
Apart from the error messages, would we break any queries by removing this check?

If not, what are some alternatives for getting good error messages?

Reply:
Looking into this more, we should just have a `else { shouldRecrusivePlan = true; }` here and reword the error messages.

Otherwise we get into a world of trouble with things like unions in subqueries. 
https://github.com/citusdata/citus/tree/unions